cloudify/decorators.py

```diff
@@ -40,4 +45,87 @@ def workflow(func=None, system_wide=False, resumable=False, **kwargs):
                                    resumable=resumable, **kwargs)
 
 
+def serial_operation(revert_to_parallel_threshold=0,
+                     workflows=None,
+                     states=None):
+    """Control order of multiple node instances of a single node template.
+        Normally, multiple node instance operations will be
+        executed in parallel.
+        However, some cases, like clusters require that cluster members are
+        installed serially.
+
+    :param revert_to_parallel_threshold: Switch back to parallel execution
+        after n node instances. Integer.
+        0: Never use parallel. Always straight serial.
+        1: After 1 node is installed, return to Parallel. Default.
+        n + 1: After n node instances.
+    :param workflows: List of workflows to apply serialization
+        to. The default is ['install']. This means we scale will
+        execute in parallel.
+    :param states: List of states that the preceding
+        node instance may be in before executing the current node instance.
+        The default is ['started']. We will only continue to next instance
+        after the preceding instance is in "started" state.
+    :raises: CloudifySerializationRetry.
+    :return: decorator wrapper
+    """
+
+    workflows = workflows if isinstance(
+        workflows, list) else ['install']
+    states = states if isinstance(
+        states, list) else ['started']
+    rest_client = get_rest_client()
+
+    def outer_wrapper(func):
+        def inner_wrapper(*args, **kwargs):
+            _ctx = kwargs.get('ctx')
+            if _ctx.workflow_id in workflows:
+                # Split the instances at the index into
+                # "priority_instances" and "no_priority"
+                # node instances. "priority_instances means that these
+                # instances' index is a smaller number than the
+                # current node instances's. They should be executed first.
+                # "no_priority" means that it comes after.
+                priority_instances, no_priority = \
+                    split_instances_at_index(rest_client)
+                # Finished instances are those that are already in the
+                # state where they are not considered blockers of our
+                # current node instance.
+                non_blocking_instances = [
+                    i for i in priority_instances + no_priority
+                    if i.state in states]
```

Comment:
i'd refactor this to return early right here, if we already hit the threshold, because then we would be able to forget about the threshold after this, and the big `if` condition would've been less crazy
(which would also imply factoring the logic out to a separate `def should_wait` function)

Reply:
will do